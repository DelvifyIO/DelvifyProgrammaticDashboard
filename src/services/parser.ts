import request from 'request';
import csv from 'csv-parser';

export function parseFromUrl(url): Promise<object[]> {
  return new Promise((resolve, reject) => {
    const results = [];
    request(url)
      .pipe(csv())
      .on('data', (data) => {
        results.push(data);
      })
      .on('end', () => resolve(results))
      .on('error', (err) => reject(err));
  });
}
