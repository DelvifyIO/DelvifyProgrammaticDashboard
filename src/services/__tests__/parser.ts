import { parseFromUrl } from '../parser';

describe('parseFromUrl', () => {
  it('parses from url', async () => {
    const url = 'https://docs.google.com/uc?export=download&id=1q2WkwkL3adsMntZiGW9RNA1ldARDGJYM';
    const result = await parseFromUrl(url);
    expect(result).toMatchSnapshot();
  });
});
