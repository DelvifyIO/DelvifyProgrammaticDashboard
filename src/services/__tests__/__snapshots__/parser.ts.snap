// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseFromUrl parses from url 1`] = `
Object {
  "data": Array [
    Object {
      "Advertiser Currency": "HKD",
      "Campaign": "HK_Columbus_DBS - Black Card",
      "Campaign ID": "3648110",
      "Click Rate (CTR)": "0.024%",
      "Clicks": "456",
      "Impressions": "1923953",
      "Insertion Order": "HK_Columbus_DBS Black Card Prospecting",
      "Insertion Order ID": "11953948",
      "Revenue (Adv Currency)": "5502.576159",
      "Revenue eCPA (Adv Currency)": "13.101372",
      "Revenue eCPC (Adv Currency)": "12.067053",
      "Revenue eCPM (Adv Currency)": "2.860037",
      "Total Conversions": "420.00",
    },
    Object {
      "Advertiser Currency": "HKD",
      "Campaign": "HK_Columbus_DBS - Black Card",
      "Campaign ID": "3648110",
      "Click Rate (CTR)": "0.097%",
      "Clicks": "842",
      "Impressions": "865233",
      "Insertion Order": "HK_Columbus_DBS Black Card Retargeting",
      "Insertion Order ID": "11953966",
      "Revenue (Adv Currency)": "9438.426401",
      "Revenue eCPA (Adv Currency)": "726.032800",
      "Revenue eCPC (Adv Currency)": "11.209533",
      "Revenue eCPM (Adv Currency)": "10.908537",
      "Total Conversions": "13.00",
    },
    Object {
      "Advertiser Currency": "HKD",
      "Campaign": "HK_Columbus_DBS - Black Card",
      "Campaign ID": "3648110",
      "Click Rate (CTR)": "0.662%",
      "Clicks": "413",
      "Impressions": "62340",
      "Insertion Order": "HK_Columbus_DBS Black Card (Outcome IO)",
      "Insertion Order ID": "12150345",
      "Revenue (Adv Currency)": "1011.630027",
      "Revenue eCPA (Adv Currency)": "1011.630027",
      "Revenue eCPC (Adv Currency)": "2.449467",
      "Revenue eCPM (Adv Currency)": "16.227623",
      "Total Conversions": "1.00",
    },
    Object {
      "Advertiser Currency": "HKD",
      "Campaign": "HK_Columbus_DBS - Loans",
      "Campaign ID": "3664042",
      "Click Rate (CTR)": "0.037%",
      "Clicks": "108",
      "Impressions": "289683",
      "Insertion Order": "HK_Columbus_DBS Loan - Tax Loan",
      "Insertion Order ID": "12017263",
      "Revenue (Adv Currency)": "3791.970378",
      "Revenue eCPA (Adv Currency)": "1895.985189",
      "Revenue eCPC (Adv Currency)": "35.110837",
      "Revenue eCPM (Adv Currency)": "13.090069",
      "Total Conversions": "2.00",
    },
    Object {
      "Advertiser Currency": "HKD",
      "Campaign": "HK_Columbus_DBS - Loans",
      "Campaign ID": "3664042",
      "Click Rate (CTR)": "0.168%",
      "Clicks": "29",
      "Impressions": "17268",
      "Insertion Order": "HK_Columbus_DBS Loan - Debt Consolidation",
      "Insertion Order ID": "12021346",
      "Revenue (Adv Currency)": "261.384464",
      "Revenue eCPA (Adv Currency)": "0.000000",
      "Revenue eCPC (Adv Currency)": "9.013257",
      "Revenue eCPM (Adv Currency)": "15.136927",
      "Total Conversions": "0.00",
    },
    Object {
      "Advertiser Currency": "HKD",
      "Campaign": "HK_Columbus_DBS - Loans",
      "Campaign ID": "3664042",
      "Click Rate (CTR)": "0.000%",
      "Clicks": "0",
      "Impressions": "242",
      "Insertion Order": "HK_Columbus_DBS Loan - Personal Loan",
      "Insertion Order ID": "12021347",
      "Revenue (Adv Currency)": "4.212749",
      "Revenue eCPA (Adv Currency)": "0.000000",
      "Revenue eCPC (Adv Currency)": "0.000000",
      "Revenue eCPM (Adv Currency)": "17.408055",
      "Total Conversions": "0.00",
    },
    Object {
      "Advertiser Currency": "HKD",
      "Campaign": "HK_Columbus_DBS - Treasures",
      "Campaign ID": "3701804",
      "Click Rate (CTR)": "0.022%",
      "Clicks": "167",
      "Impressions": "771445",
      "Insertion Order": "HK_Columbus_DBS Treasures Prospecting",
      "Insertion Order ID": "12148742",
      "Revenue (Adv Currency)": "2693.046726",
      "Revenue eCPA (Adv Currency)": "538.609345",
      "Revenue eCPC (Adv Currency)": "16.126028",
      "Revenue eCPM (Adv Currency)": "3.490912",
      "Total Conversions": "5.00",
    },
    Object {
      "Advertiser Currency": "HKD",
      "Campaign": "HK_Columbus_DBS - Treasures",
      "Campaign ID": "3701804",
      "Click Rate (CTR)": "0.182%",
      "Clicks": "311",
      "Impressions": "170881",
      "Insertion Order": "HK_Columbus_DBS Treasures Retargeting",
      "Insertion Order ID": "12148743",
      "Revenue (Adv Currency)": "2458.361034",
      "Revenue eCPA (Adv Currency)": "74.495789",
      "Revenue eCPC (Adv Currency)": "7.904698",
      "Revenue eCPM (Adv Currency)": "14.386392",
      "Total Conversions": "33.00",
    },
    Object {
      "Advertiser Currency": "HKD",
      "Campaign": "HK_Columbus_DBS - Treasures",
      "Campaign ID": "3701804",
      "Click Rate (CTR)": "0.464%",
      "Clicks": "172",
      "Impressions": "37091",
      "Insertion Order": "HK_Columbus_DBS Treasures Outcome Buying",
      "Insertion Order ID": "12150340",
      "Revenue (Adv Currency)": "1005.600027",
      "Revenue eCPA (Adv Currency)": "502.800013",
      "Revenue eCPC (Adv Currency)": "5.846512",
      "Revenue eCPM (Adv Currency)": "27.111699",
      "Total Conversions": "2.00",
    },
    Object {
      "Advertiser Currency": "HKD",
      "Campaign": "HK_Columbus_DBS - Treasures",
      "Campaign ID": "3701804",
      "Click Rate (CTR)": "0.024%",
      "Clicks": "169",
      "Impressions": "708430",
      "Insertion Order": "HK_Columbus_DBS Treasures Online Prospecting",
      "Insertion Order ID": "12201146",
      "Revenue (Adv Currency)": "2678.191095",
      "Revenue eCPA (Adv Currency)": "243.471918",
      "Revenue eCPC (Adv Currency)": "15.847285",
      "Revenue eCPM (Adv Currency)": "3.780460",
      "Total Conversions": "11.00",
    },
  ],
  "totals": Object {
    "Click Rate (CTR)": "0.055%",
    "Clicks": "2667",
    "Impressions": "4846566",
    "Revenue (Adv Currency)": "28845.399060",
    "Revenue eCPA (Adv Currency)": "59.230799",
    "Revenue eCPC (Adv Currency)": "10.815673",
    "Revenue eCPM (Adv Currency)": "5.951719",
    "Total Conversions": "487.00",
  },
}
`;
